File: ddns/provider/alidns.py
# coding=utf-8
"""
AliDNS API
阿里DNS解析操作库
@author: NewFuture
"""

from time import gmtime, strftime, time

from ._base import TYPE_FORM, BaseProvider, encode_params, join_domain
from ._signature import hmac_sha256_authorization, sha256_hash


class AliBaseProvider(BaseProvider):
    """阿里云基础Provider，提供通用的_request方法"""

    endpoint = "https://alidns.aliyuncs.com"
    content_type = TYPE_FORM  # 阿里云DNS API使用表单格式
    api_version = "2015-01-09"  # API版本，v3签名需要

    def _request(self, action, method="POST", **params):
        # type: (str, str, **(Any)) -> dict
        """Aliyun v3 https://help.aliyun.com/zh/sdk/product-overview/v3-request-structure-and-signature"""
        params = {k: v for k, v in params.items() if v is not None}

        if method in ("GET", "DELETE"):
            # For GET and DELETE requests, parameters go in query string
            query_string = encode_params(params) if len(params) > 0 else ""
            body_content = ""
        else:
            # For POST requests, parameters go in body
            body_content = self._encode_body(params)
            query_string = ""

        path = "/"
        content_hash = sha256_hash(body_content)
        # 构造请求头部
        headers = {
            "host": self.endpoint.split("://", 1)[1].strip("/"),
            "content-type": self.content_type,
            "x-acs-action": action,
            "x-acs-content-sha256": content_hash,
            "x-acs-date": strftime("%Y-%m-%dT%H:%M:%SZ", gmtime()),
            "x-acs-signature-nonce": str(hash(time()))[2:],
            "x-acs-version": self.api_version,
        }

        # 使用通用签名函数
        authorization = hmac_sha256_authorization(
            secret_key=self.token,
            method=method,
            path=path,
            query=query_string,
            headers=headers,
            body_hash=content_hash,
            signing_string_format="ACS3-HMAC-SHA256\n{HashedCanonicalRequest}",
            authorization_format=(
                "ACS3-HMAC-SHA256 Credential=" + self.id + ",SignedHeaders={SignedHeaders},Signature={Signature}"
            ),
        )
        headers["Authorization"] = authorization
        # 对于v3签名的RPC API，参数在request body中
        path = path if not query_string else path + "?" + format(query_string)
        return self._http(method, path, body=body_content, headers=headers)


class AlidnsProvider(AliBaseProvider):
    """阿里云DNS Provider"""

    def _split_zone_and_sub(self, domain):
        # type: (str) -> tuple[str | None, str | None, str]
        """
        AliDNS 支持直接查询主域名和RR，无需循环查询。
        返回没有DomainId,用DomainName代替
        https://help.aliyun.com/zh/dns/api-alidns-2015-01-09-getmaindomainname
        """
        res = self._request("GetMainDomainName", InputString=domain)
        sub, main = res.get("RR"), res.get("DomainName")
        return (main, sub, main or domain)

    def _query_zone_id(self, domain):
        """调用_split_zone_and_sub可直接获取，无需调用_query_zone_id"""
        raise NotImplementedError("_split_zone_and_sub is used to get zone_id")

    def _query_record(self, zone_id, subdomain, main_domain, record_type, line, extra):
        """https://help.aliyun.com/zh/dns/api-alidns-2015-01-09-describesubdomainrecords"""
        sub = join_domain(subdomain, main_domain)
        data = self._request(
            "DescribeSubDomainRecords",
            SubDomain=sub,  # aliyun API要求SubDomain为完整域名
            DomainName=main_domain,
            Type=record_type,
            Line=line,
            PageSize=500,
            Lang=extra.get("Lang"),  # 默认中文
            Status=extra.get("Status"),  # 默认全部状态
        )
        records = data.get("DomainRecords", {}).get("Record", [])
        if not records:
            self.logger.warning(
                "No records found for [%s] with %s <%s> (line: %s)", zone_id, subdomain, record_type, line
            )
        elif not isinstance(records, list):
            self.logger.error("Invalid records format: %s", records)
        else:
            return next((r for r in records), None)
        return None

    def _create_record(self, zone_id, subdomain, main_domain, value, record_type, ttl, line, extra):
        """https://help.aliyun.com/zh/dns/api-alidns-2015-01-09-adddomainrecord"""
        data = self._request(
            "AddDomainRecord",
            DomainName=main_domain,
            RR=subdomain,
            Value=value,
            Type=record_type,
            TTL=ttl,
            Line=line,
            **extra
        )  # fmt: skip
        if data and data.get("RecordId"):
            self.logger.info("Record created: %s", data)
            return True
        self.logger.error("Failed to create record: %s", data)
        return False

    def _update_record(self, zone_id, old_record, value, record_type, ttl, line, extra):
        """https://help.aliyun.com/zh/dns/api-alidns-2015-01-09-updatedomainrecord"""
        # 阿里云DNS update新旧值不能一样，先判断是否发生变化
        if (
            old_record.get("Value") == value
            and old_record.get("Type") == record_type
            and (not ttl or old_record.get("TTL") == ttl)
        ):
            domain = join_domain(old_record.get("RR"), old_record.get("DomainName"))
            self.logger.warning("No changes detected, skipping update for record: %s", domain)
            return True
        data = self._request(
            "UpdateDomainRecord",
            RecordId=old_record.get("RecordId"),
            Value=value,
            RR=old_record.get("RR"),
            Type=record_type,
            TTL=ttl,
            Line=line or old_record.get("Line"),
            **extra
        )  # fmt: skip
        if data and data.get("RecordId"):
            self.logger.info("Record updated: %s", data)
            return True
        self.logger.error("Failed to update record: %s", data)
        return False

File: rust/src/provider/alidns.rs
use std::collections::BTreeMap;

use serde_json::Value;

use crate::error::{Error, Result};
use crate::http::{Method, form_encode};
use crate::signature::{acs_timestamp, acs3_authorization, request_nonce, sha256_hex};

use super::base::{
    CrudProvider, ProviderContext, RecordRequest, ZoneMatch, endpoint_host, join_domain,
    string_parameters, value_to_string,
};

pub struct AlidnsProvider {
    context: ProviderContext,
    zones: BTreeMap<String, String>,
}

impl AlidnsProvider {
    pub fn new(context: ProviderContext) -> Result<Self> {
        if context.id.is_empty() {
            return Err(Error::Config(
                "AliDNS access key id must be configured".to_owned(),
            ));
        }
        if context.token.is_empty() {
            return Err(Error::Config(
                "AliDNS access key secret must be configured".to_owned(),
            ));
        }
        Ok(Self {
            context,
            zones: BTreeMap::new(),
        })
    }

    fn api(&self, action: &str, parameters: BTreeMap<String, String>) -> Result<Value> {
        let body = form_encode(&parameters);
        let body_hash = sha256_hex(&body);
        let mut headers = BTreeMap::from([
            (
                "host".to_owned(),
                endpoint_host(&self.context.endpoint, "AliDNS")?,
            ),
            (
                "content-type".to_owned(),
                "application/x-www-form-urlencoded".to_owned(),
            ),
            ("x-acs-action".to_owned(), action.to_owned()),
            ("x-acs-content-sha256".to_owned(), body_hash.clone()),
            ("x-acs-date".to_owned(), acs_timestamp()),
            ("x-acs-signature-nonce".to_owned(), request_nonce()?),
            ("x-acs-version".to_owned(), "2015-01-09".to_owned()),
        ]);
        let authorization = acs3_authorization(
            &self.context.id,
            &self.context.token,
            "POST",
            "/",
            "",
            &headers,
            &body_hash,
        )?;
        headers.insert("authorization".to_owned(), authorization);
        let response =
            self.context
                .send_json(Method::Post, "/", &BTreeMap::new(), Some(body), headers)?;
        if let Some(code) = response.get("Code").and_then(Value::as_str) {
            let message = response
                .get("Message")
                .and_then(Value::as_str)
                .unwrap_or("unknown error");
            return Err(Error::Provider(format!(
                "AliDNS API error {code}: {}",
                self.context.logger.mask(message)
            )));
        }
        Ok(response)
    }
}

impl CrudProvider for AlidnsProvider {
    fn name(&self) -> &'static str {
        "alidns"
    }

    fn context(&self) -> &ProviderContext {
        &self.context
    }

    fn zone_cache(&mut self) -> &mut BTreeMap<String, String> {
        &mut self.zones
    }

    fn split_zone_and_sub(&mut self, domain: &str) -> Result<ZoneMatch> {
        if let Some((subdomain, main_domain)) = super::base::split_custom_domain(domain) {
            return Ok(ZoneMatch {
                zone_id: main_domain.clone(),
                subdomain,
                main_domain,
            });
        }
        let response = self.api(
            "GetMainDomainName",
            BTreeMap::from([("InputString".to_owned(), domain.to_owned())]),
        )?;
        let main_domain = response
            .get("DomainName")
            .and_then(Value::as_str)
            .ok_or_else(|| Error::Provider(format!("AliDNS could not split domain `{domain}`")))?
            .to_owned();
        let subdomain = response
            .get("RR")
            .and_then(Value::as_str)
            .unwrap_or("@")
            .to_owned();
        Ok(ZoneMatch {
            zone_id: main_domain.clone(),
            subdomain,
            main_domain,
        })
    }

    fn query_zone_id(&mut self, _domain: &str) -> Result<Option<String>> {
        Err(Error::Provider(
            "AliDNS obtains the main domain through GetMainDomainName".to_owned(),
        ))
    }

    fn query_record(
        &mut self,
        zone_id: &str,
        subdomain: &str,
        main_domain: &str,
        request: &RecordRequest,
    ) -> Result<Option<Value>> {
        let parameters = [
            ("SubDomain", Some(join_domain(subdomain, main_domain))),
            ("DomainName", Some(main_domain.to_owned())),
            ("Type", Some(request.record_type.clone())),
            ("Line", request.line.clone()),
            ("PageSize", Some("500".to_owned())),
            ("Lang", request.extra.get("Lang").and_then(value_to_string)),
            (
                "Status",
                request.extra.get("Status").and_then(value_to_string),
            ),
        ]
        .into_iter()
        .filter_map(|(key, value)| value.map(|value| (key.to_owned(), value)))
        .collect();
        let response = self.api("DescribeSubDomainRecords", parameters)?;
        let record = response
            .pointer("/DomainRecords/Record")
            .and_then(Value::as_array)
            .and_then(|records| records.first())
            .cloned();
        if record.is_none() {
            self.context.logger.info(
                "alidns",
                format!("no {} record found in zone {zone_id}", request.record_type),
            );
        }
        Ok(record)
    }

    fn create_record(
        &mut self,
        _zone_id: &str,
        subdomain: &str,
        main_domain: &str,
        request: &RecordRequest,
    ) -> Result<()> {
        let parameters = string_parameters(
            request,
            [
                ("DomainName", Some(main_domain.to_owned())),
                ("RR", Some(subdomain.to_owned())),
                ("Value", Some(request.address.clone())),
                ("Type", Some(request.record_type.clone())),
                ("TTL", request.ttl.map(|ttl| ttl.to_string())),
                ("Line", request.line.clone()),
            ],
        );
        let response = self.api("AddDomainRecord", parameters)?;
        if response.get("RecordId").is_some() {
            Ok(())
        } else {
            Err(Error::Provider(format!(
                "AliDNS failed to create record: {response}"
            )))
        }
    }

    fn update_record(
        &mut self,
        _zone_id: &str,
        record: &Value,
        request: &RecordRequest,
    ) -> Result<()> {
        let unchanged = record.get("Value").and_then(Value::as_str)
            == Some(request.address.as_str())
            && record.get("Type").and_then(Value::as_str) == Some(request.record_type.as_str())
            && request.ttl.is_none_or(|ttl| {
                record
                    .get("TTL")
                    .and_then(Value::as_u64)
                    .is_some_and(|old| old == u64::from(ttl))
            });
        if unchanged {
            self.context
                .logger
                .info("alidns", "record already has the requested value");
            return Ok(());
        }
        let parameters = string_parameters(
            request,
            [
                ("RecordId", record.get("RecordId").and_then(value_to_string)),
                ("Value", Some(request.address.clone())),
                ("RR", record.get("RR").and_then(value_to_string)),
                ("Type", Some(request.record_type.clone())),
                ("TTL", request.ttl.map(|ttl| ttl.to_string())),
                (
                    "Line",
                    request
                        .line
                        .clone()
                        .or_else(|| record.get("Line").and_then(value_to_string)),
                ),
            ],
        );
        let response = self.api("UpdateDomainRecord", parameters)?;
        if response.get("RecordId").is_some() {
            Ok(())
        } else {
            Err(Error::Provider(format!(
                "AliDNS failed to update record: {response}"
            )))
        }
    }
}
